diff -urN libusb-0.1.0.org/usb.h libusb-0.1.0/usb.h
--- libusb-0.1.0.org/usb.h	Sun Aug  5 21:37:18 2001
+++ libusb-0.1.0/usb.h	Sun Aug  5 21:37:52 2001
@@ -203,5 +203,5 @@
 void usb_init(void);
 struct usb_device *usb_device(usb_dev_handle *dev);
 
-#endif __USB_H__
+#endif /* __USB_H__ */
 
diff -urN libusb-0.1.0.org/usbi.h libusb-0.1.0/usbi.h
--- libusb-0.1.0.org/usbi.h	Sun Aug  5 21:37:18 2001
+++ libusb-0.1.0/usbi.h	Sun Aug  5 21:38:04 2001
@@ -16,5 +16,5 @@
   int altsetting;
 };
 
-#endif __USBI_H__
+#endif /* __USBI_H__ */
 
